packages/plugin-ext/src/main/browser/tabs/tabs-main.ts

```diff
@@ -15,28 +15,301 @@
 // *****************************************************************************
 
 import { interfaces } from '@theia/core/shared/inversify';
-
-import { TabsMain } from '../../../common/plugin-api-rpc';
+import { ApplicationShell, PINNED_CLASS, Saveable, TabBar, Title, ViewContainer, Widget } from '@theia/core/lib/browser';
+import { AnyInputDto, MAIN_RPC_CONTEXT, TabDto, TabGroupDto, TabInputKind, TabModelOperationKind, TabsExt, TabsMain } from '../../../common/plugin-api-rpc';
 import { RPCProtocol } from '../../../common/rpc-protocol';
+import { EditorPreviewWidget } from '@theia/editor-preview/lib/browser/editor-preview-widget';
+import { Disposable } from '@theia/core/shared/vscode-languageserver-protocol';
+import { MonacoDiffEditor } from '@theia/monaco/lib/browser/monaco-diff-editor';
+import { toUriComponents } from '../hierarchy/hierarchy-types-converters';
+import { TerminalWidget } from '@theia/terminal/lib/browser/base/terminal-widget';
+
+interface TabInfo {
+    tab: TabDto;
+    tabIndex: number;
+    group: TabGroupDto;
+}
+
+export class TabsMainImpl implements TabsMain, Disposable {
+
+    private readonly proxy: TabsExt;
+    private tabGroupModel = new Map<TabBar<Widget>, TabGroupDto>();
+    private tabInfoLookup = new Map<Title<Widget>, TabInfo>();
+
+    private applicationShell: ApplicationShell;
+
+    private disposableTabBarListeners: Disposable[] = [];
+    private toDisposeOnDestroy: Disposable[] = [];
 
-export class TabsMainImp implements TabsMain {
+    private groupIdCounter = 0;
+    private currentActiveGroup: TabGroupDto;
+
+    private tabGroupChanged: boolean = false;
 
     constructor(
         rpc: RPCProtocol,
         container: interfaces.Container
-    ) {}
+    ) {
+        this.proxy = rpc.getProxy(MAIN_RPC_CONTEXT.TABS_EXT);
+
+        this.applicationShell = container.get(ApplicationShell);
+        this.createTabsModel();
+
+        const tabBars = this.applicationShell.mainPanel.tabBars();
+        for (let tabBar; tabBar = tabBars.next();) {
+            this.attachListenersToTabBar(tabBar);
+        }
+
+        this.toDisposeOnDestroy.push(
+            this.applicationShell.mainPanelRenderer.onDidCreateTabBar(tabBar => {
+                this.attachListenersToTabBar(tabBar);
+                this.onTabGroupCreated(tabBar);
+            })
+        );
+
+        this.connectToSignal(this.toDisposeOnDestroy, this.applicationShell.mainPanel.widgetAdded, (mainPanel, widget) => {
+            if (this.tabGroupChanged || this.tabGroupModel.size === 0) {
+                this.tabGroupChanged = false;
+                this.createTabsModel();
+            } else {
+                const tabBar = mainPanel.findTabBar(widget.title)!;
+                const oldTabInfo = this.tabInfoLookup.get(widget.title);
```

Comment:
The implication of the `tabInforLookup` is that titles must be unique across all tab groups. Does the API or the Theia implementation of tabs imply or mandate that?

Reply:
The Tabs instances are coming from phosphor. 
I tried opening the same tab multiple times and the title objects are still different. So it seems each tab definatly has its own title object